File: Go/ExampleTests/common/framedisplay_string.go
package common

type FrameDisplayString struct {
	Frame string
	Mark1 string
	Mark2 string
	Mark3 string
	TotalScore string
}

func NewFrameDisplayStringFromSlice(v []string) FrameDisplayString {
	s := FrameDisplayString{}
	if len(v) > 0 { s.Frame = v[0] }
	if len(v) > 1 { s.Mark1 = v[1] }
	if len(v) > 2 { s.Mark2 = v[2] }
	if len(v) > 3 { s.Mark3 = v[3] }
	if len(v) > 4 { s.TotalScore = v[4] }
	return s
}

// NewFrameDisplayStringFromText builds from the text form, e.g. Money as "25 USD".
func NewFrameDisplayStringFromText(text string) FrameDisplayString {
	parts := RequireTokens(text, 5, "FrameDisplay")
	return FrameDisplayString{
		Frame: parts[0],
		Mark1: parts[1],
		Mark2: parts[2],
		Mark3: parts[3],
		TotalScore: parts[4],
	}
}

func (s FrameDisplayString) String() string {
	return Token(s.Frame) + " " + Token(s.Mark1) + " " + Token(s.Mark2) + " " + Token(s.Mark3) + " " + Token(s.TotalScore)
}

func (s FrameDisplayString) Equals(o FrameDisplayString) bool {
	return DNCEqual(s.Frame, o.Frame) &&
		DNCEqual(s.Mark1, o.Mark1) &&
		DNCEqual(s.Mark2, o.Mark2) &&
		DNCEqual(s.Mark3, o.Mark3) &&
		DNCEqual(s.TotalScore, o.TotalScore)
}

func EqualFrameDisplayStringSlices(a, b []FrameDisplayString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}

File: Rust/ExampleTests/common/framedisplay_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

#[derive(Debug, Clone, Default)]
pub struct FrameDisplayString {
    pub frame: String,
    pub mark1: String,
    pub mark2: String,
    pub mark3: String,
    pub totalscore: String,
}

impl FrameDisplayString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            frame: v.get(0).copied().unwrap_or("").to_string(),
            mark1: v.get(1).copied().unwrap_or("").to_string(),
            mark2: v.get(2).copied().unwrap_or("").to_string(),
            mark3: v.get(3).copied().unwrap_or("").to_string(),
            totalscore: v.get(4).copied().unwrap_or("").to_string(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 5, "FrameDisplay");
        Self {
            frame: parts[0].clone(),
            mark1: parts[1].clone(),
            mark2: parts[2].clone(),
            mark3: parts[3].clone(),
            totalscore: parts[4].clone(),
        }
    }
}

impl std::fmt::Display for FrameDisplayString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.frame), crate::common::tokens::token(&self.mark1), crate::common::tokens::token(&self.mark2), crate::common::tokens::token(&self.mark3), crate::common::tokens::token(&self.totalscore)].join(" "))
    }
}

impl PartialEq for FrameDisplayString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.frame, &other.frame)
            && crate::common::dnc_equal(&self.mark1, &other.mark1)
            && crate::common::dnc_equal(&self.mark2, &other.mark2)
            && crate::common::dnc_equal(&self.mark3, &other.mark3)
            && crate::common::dnc_equal(&self.totalscore, &other.totalscore)
    }
}
